### PR TITLE
REFACTOR-#2011: move default_to_pandas in groupby to backend

### LINKED ISSUE #2011: [REFACTOR] Don't call default_to_pandas in API level for groupby by a mix of column names and series
We need it for NYC Taxi Q3

### PR DESCRIPTION
**Pull request by ienkovich**
## What do these changes do?
In OmniSci backend we can handle some cases of `grouby` where `by` holds both column names and series. We need this feature for NYC Taxi benchmark. To utilize backend support we need to move `default_to_pandas` call from API level to the backend.

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #2011  
- [x] tests passing

### CODE REVIEW

File: modin/data_management/functions/groupby_function.py [hunk 1]
```diff
@@ -29,9 +30,13 @@ def caller(
             numeric_only=True,
             drop=False,
         ):
-            assert isinstance(
-                by, type(query_compiler)
-            ), "Can only use groupby reduce with another Query Compiler"
+            if not isinstance(by, type(query_compiler)):
```
Comment (YarShev) [inline modin/data_management/functions/groupby_function.py:33]: What cases this is working now for?

Comment (ienkovich) [reply modin/data_management/functions/groupby_function.py:33]: It works for all cases we couldn't handle in `DataFrame.groupby` and transform `by` to a `QueryCompiler`. The case of our interest is NYC Taxi query like the following:
```
df = df.groupby(["b", df["c"].dt.year]).size()
```

Comment (YarShev) [reply modin/data_management/functions/groupby_function.py:33]: I see, thanks!

Comment (YarShev) [reply modin/data_management/functions/groupby_function.py:33]: I see, thanks!

**Review by YarShev — APPROVED** [review]
@ienkovich , LGTM!

**Pull request merged**